src/rdwin32.h

```diff
@@ -223,37 +223,113 @@ static RD_UNUSED int rd_fd_set_nonblocking (int fd) {
  * @returns 0 on success or errno on failure
  */
 static RD_UNUSED int rd_pipe_nonblocking (int *fds) {
-        HANDLE h[2];
-        int i;
-
-        if (!CreatePipe(&h[0], &h[1], NULL, 0))
-                return (int)GetLastError();
-        for (i = 0 ; i < 2 ; i++) {
-                DWORD mode = PIPE_NOWAIT;
-                /* Set non-blocking */
-                if (!SetNamedPipeHandleState(h[i], &mode, NULL, NULL)) {
-                        CloseHandle(h[0]);
-                        CloseHandle(h[1]);
-                        return (int)GetLastError();
-                }
-
-                /* Open file descriptor for handle */
-                fds[i] = _open_osfhandle((intptr_t)h[i],
-                                         i == 0 ?
-                                         O_RDONLY | O_BINARY :
-                                         O_WRONLY | O_BINARY);
-
-                if (fds[i] == -1) {
-                        CloseHandle(h[0]);
-                        CloseHandle(h[1]);
-                        return (int)GetLastError();
-                }
+        /* On windows, the "pipe" will be a tcp connection.
+        * This is to allow WSAPoll to be used to poll pipe events */
+
+        SOCKET listen_s = INVALID_SOCKET;
+        SOCKET accept_s = INVALID_SOCKET;
+        SOCKET connect_s = INVALID_SOCKET;
+
+        WSAPOLLFD poll_fd;
+
+        struct sockaddr_in listen_addr;
+        struct sockaddr_in connect_addr;
+        socklen_t sock_len = 0;
+
+        int poll_timeout_ms = 1000;
+
+        /* Create listen socket */
+        listen_s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+        if (listen_s == INVALID_SOCKET)
+                goto err;
+
+        if (rd_fd_set_nonblocking(listen_s) != 0)
+                goto err;
+
+        listen_addr.sin_family = AF_INET;
+        listen_addr.sin_addr.s_addr = ntohl(INADDR_ANY);
+        listen_addr.sin_port = 0;
+        if (bind(listen_s, (struct sockaddr*)&listen_addr, sizeof(struct sockaddr_in)) != 0)
+                goto err;
+
+        sock_len = sizeof(struct sockaddr_in);
+        if (getsockname(listen_s, (struct sockaddr*)&connect_addr, &sock_len) != 0)
+                goto err;
+
+        if (listen(listen_s, 1) != 0)
+                goto err;
+
+        /* Create connection socket */
+        connect_s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+        if (connect_s == INVALID_SOCKET)
+                goto err;
+
+        if (rd_fd_set_nonblocking(connect_s) != 0)
+                goto err;
+
+        connect_addr.sin_addr.s_addr = ntohl(INADDR_LOOPBACK);
+        if (connect(
+                connect_s,
+                (struct sockaddr*)&connect_addr,
+                sizeof(struct sockaddr_in)) == SOCKET_ERROR) {
+                if (WSAGetLastError() != WSAEWOULDBLOCK)
+                        goto err;
         }
+
+        /* Wait for incoming connection */
```

Comment:
Do you really need the poll here? Can't it just block on the accept?

Reply:
It needs to be polled due to both the listen and connection sockets being non-blocking. The alternative is to launch a thread to initiate the connection.

Reply:
The listen socket could be made blocking, but I used the polling method to guard against any chance of hanging on accept.

Reply:
Can't you set it to non-blocking *after* the connection has been established?

Reply:
In the most recent change, the listen socket is blocking and the returned socket from accept is set to non-blocking. The connection socket must be non-blocking to connect in a single threaded manner since the call to connect would block otherwise.